src/abb_system_position_only.cpp

```diff
@@ -13,242 +13,200 @@
+      << abb::robot::summaryText(robot_controller_description_));
 
   for (const hardware_interface::ComponentInfo & joint : info_.joints) {
     if (joint.command_interfaces.size() != 1) {
-      RCLCPP_FATAL(
-        rclcpp::get_logger("ABBSystemPositionOnlyHardware"),
-        "expecting exactly 1 command interface");
+      RCLCPP_FATAL(LOGGER, "Expecting exactly 1 command interface");
       return CallbackReturn::ERROR;
     }
 
     if (joint.command_interfaces[0].name != hardware_interface::HW_IF_POSITION) {
-      RCLCPP_FATAL(
-        rclcpp::get_logger("ABBSystemPositionOnlyHardware"),
-        "expecting only POSITION command interface");
+      RCLCPP_FATAL(LOGGER, "Expecting only POSITION command interface");
       return CallbackReturn::ERROR;
     }
 
     if (joint.state_interfaces.size() != 1) {
-      RCLCPP_FATAL(
-        rclcpp::get_logger("ABBSystemPositionOnlyHardware"), "expecting exactly 1 state interface");
+      RCLCPP_FATAL(LOGGER, "Expecting exactly 1 state interface");
       return CallbackReturn::ERROR;
     }
 
     if (joint.state_interfaces[0].name != hardware_interface::HW_IF_POSITION) {
-      RCLCPP_FATAL(
-        rclcpp::get_logger("ABBSystemPositionOnlyHardware"),
-        "expecting only POSITION state interface");
+      RCLCPP_FATAL(LOGGER, "Expecting only POSITION state interface");
       return CallbackReturn::ERROR;
     }
   }
 
   // Configure EGM
-  RCLCPP_INFO(rclcpp::get_logger("ABBSystemPositionOnlyHardware"), "Configuring EGM interface...");
-
-  egm_interface_ =
-    std::make_unique<abb::egm::EGMControllerInterface>(io_service_, robotstudio_port_);
-  if (!egm_interface_) {
-    RCLCPP_FATAL(
-      rclcpp::get_logger("ABBSystemPositionOnlyHardware"),
-      "Could not create EGMControllerInterface");
+  RCLCPP_INFO(LOGGER, "Configuring EGM interface...");
+
+  // Initialize motion data from robot controller description
+  try {
+    abb::robot::initializeMotionData(motion_data_, robot_controller_description_);
+  } catch (...) {
+    RCLCPP_ERROR_STREAM(
+      LOGGER, "Failed to initialize motion data from robot controller description");
     return CallbackReturn::ERROR;
   }
-
-  if (!egm_interface_->isInitialized()) {
-    RCLCPP_FATAL(
-      rclcpp::get_logger("ABBSystemPositionOnlyHardware"),
-      "EGM UDP Server interface failed to initialize (e.g. due to port already bound)");
+  
+  // Create channel configuration for each mechanical unit group
+  std::vector<abb::robot::EGMManager::ChannelConfiguration> channel_configurations;
+  for (const auto & group : robot_controller_description_.mechanical_units_groups()) {
+    // Each channel needs a new port number, start with port from hardware parameters and increment
+    // by 1 for each channel
+    auto channel_configuration =
+      abb::robot::EGMManager::ChannelConfiguration{static_cast<uint16_t>(robotstudio_port), group};
+    channel_configurations.emplace_back(channel_configuration);
+    robotstudio_port += 1;
```

Comment:
I couldn't find a way to get the remote port number for each mechanical unit group. For now, we can perhaps assume sequentially-increasing port numbers, but I hope there is a more correct way to do this.

Reply:
This depends on the EGM configuration on the controller.

This is a bit of information your user would have to provide.